pkg/object/gatewaycontroller/gatewaycontroller.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package gateway implements k8s gateway API.
+package gatewaycontroller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/megaease/easegress/v2/pkg/logger"
+	"github.com/megaease/easegress/v2/pkg/object/trafficcontroller"
+	"github.com/megaease/easegress/v2/pkg/supervisor"
+)
+
+const (
+	// Category is the name of GatewayController
+	Category = supervisor.CategoryBusinessController
+
+	// Kind is the kind of GatewayController
+	Kind = "GatewayController"
+
+	gatewayControllerName = "megaease.com/gateway-controller"
+)
+
+type (
+	// GatewayController implements a k8s gateway API controller
+	GatewayController struct {
+		super     *supervisor.Supervisor
+		superSpec *supervisor.Spec
+		spec      *Spec
+
+		tc        *trafficcontroller.TrafficController
+		namespace string
+		k8sClient *k8sClient
+
+		stopCh chan struct{}
+		wg     sync.WaitGroup
+	}
+
+	// Spec is the ingress controller spec
+	Spec struct {
+		KubeConfig string   `json:"kubeConfig" jsonschema:"omitempty"`
+		MasterURL  string   `json:"masterURL" jsonschema:"omitempty"`
+		Namespaces []string `json:"namespaces" jsonschema:"omitempty"`
+	}
+)
+
+func init() {
+	supervisor.Register(&GatewayController{})
+}
+
+// Category returns the category of GatewayController.
+func (gc *GatewayController) Category() supervisor.ObjectCategory {
+	return Category
+}
+
+// Kind returns the kind of GatewayController.
+func (gc *GatewayController) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of GatewayController.
+func (gc *GatewayController) DefaultSpec() interface{} {
+	return &Spec{}
```

Comment:
IMO, `~/.kube/config` could be the default value. And I think the GatewayController could support ClientSetInCluster mode[1].

[1] https://github.com/kubernetes/client-go/blob/master/examples/in-cluster-client-configuration/main.go

Reply:
the default is in cluster mode(when both `KubeConfig` and `MasterURL` are empty) as it will be deployed in a k8s cluster at most of the time. And ClientSetInCluster mode is used, please check L161 in k8s.go.